default/AI/ResearchAI.py

```diff
@@ -1,30 +1,328 @@
 import freeOrionAIInterface as fo  # pylint: disable=import-error
 import FreeOrionAI as foAI
-import TechsListsAI
 import AIDependencies
 import AIstate
 import traceback
 import ColonisationAI
 from freeorion_tools import tech_is_complete, chat_human
 
-inProgressTechs = {}
-
-
-def get_research_index():
-    empire_id = fo.empireID()
-    research_index = empire_id % 2
-    if foAI.foAIstate.aggression >= fo.aggression.aggressive:  # maniacal
-        research_index = 2 + (empire_id % 3)  # so indices [2,3,4]
-    elif foAI.foAIstate.aggression >= fo.aggression.typical:
-        research_index += 1
+def get_empire_hash():
+    """Get hash for empire name, so to create random tech choices in every game, yet deterministic across loading of same game
+    """
+    ename = fo.getEmpire().name
+    hash = 0
+    for char in ename:
+        hash += ord(char)
+    research_index = fo.getEmpire().empireID + hash
     return research_index
 
+def get_priority(tech_name, empire):
+    """Get tech priority. 1 is default. 0 if not useful (but doesn't hurt to research), < 0 to prevent AI to research it
+    """
 
-def exclude_tech(tech_name):
-    return ((foAI.foAIstate.aggression < AIDependencies.TECH_EXCLUSION_MAP_1.get(tech_name, fo.aggression.invalid)) or
-            (foAI.foAIstate.aggression > AIDependencies.TECH_EXCLUSION_MAP_2.get(tech_name, fo.aggression.maniacal)) or
-            tech_name in TechsListsAI.unusable_techs())
-            
+    index = get_empire_hash()
+
+    immediate = 999 # i.e. research this tech NOW!
+    useless = 0 # useless for AI, mostly because the AI doesn't know how to use them
+    default = 1
+    unresearchable = -1 # anything unresearchable should use this TODO obtain this information from techs.txt
+
+    theory = 0 # theories has 0 priority as they don't unlock anything. If it ends up prereq for any other tech those tech will add the theories in.
+
+    growth = 2 # maximum population growth has higher priority.
+    production_boost = 1.5 # production boosts have higher priority. This include buildings that boosts production
+    research_boost = 2 # research boosts have higher priority. This include buildings that boosts research
+    meter_boost = 1
+    shield = 1.25 # shields are more powerful than armors or weapons
+    defensive = 2 if foAI.foAIstate.aggression <= fo.aggression.cautious else (1 if fo.currentTurn() > 10 else 0.5)
+    supply = 2 if ColonisationAI.galaxy_is_sparse() else 0.5
```

Comment:
```ColonisationAI.galaxy_is_sparse()``` defines sparseness by the amount of players in the game compared to galaxy size (and some monsters option).
It does not consider the Planet or starlane settings which probably is what we want to consider here.